auth-service/src/main/kotlin/uk/gov/communities/delta/auth/utils/RegistrationService.kt

```diff
@@ -0,0 +1,155 @@
+import uk.gov.communities.delta.auth.config.DeltaConfig
+
+class RegistrationService(
+    private val deltaConfig: DeltaConfig,
+    private val organisationService: OrganisationService,
+    private val emailService: EmailService
+) {
+    class RegistrationResult {
+        enum class Status {
+            CREATED, ALREADY_EXISTS, OTHER_FAILURE
+        }
+
+        val status: Status
+        val email: String
+        val token: String?
+
+        constructor(
+            status: Status,
+            email: String
+        ) {
+            this.status = status
+            this.email = email
+            token = null
+        }
+
+        constructor(
+            status: Status,
+            email: String,
+            token: String
+        ) {
+            this.status = status
+            this.email = email
+            this.token = token // TODO - are these still needed with tokens stored in db (see below)
+        }
+    }
+
+    // Previously users could be datamart users but not delta users, at migration we only transferred over users who
+    // were delta users so there should never be users who exist but aren't delta users - therefore can remove all logic
+    // for attaching users and checking if they are indeed delta users (if they somehow existed they'd get a "contact
+    // the service desk" error at login which is what we want
+
+
+    private val logger = LoggerFactory.getLogger(javaClass)
+    private val userService = UserService()
+
+    //        val context = Context(Locale.getDefault(), model) TODO - what did this do? Is it still needed? - from email sending
+
+
+    fun register(registration: Registration): RegistrationResult { // TODO - does this still need to return anything?
+        val user = userService.findUserById(registration.emailAddress, true).orElse(null)
+        if (user != null) {
+            // TODO - logic check - do we want this to happen?
+            //  - if a user exists already then registering them again will add them to organisations associated with their domain
+            //      - can't see a scenario where this isn't already the case but if it is then do we definitely want to fix it?
+//            registrationOrganisationService.attachUserToDomainOrganisations(user)
+            // TODO - set this up to send an email (bring email content over from delta)
+//            deltaUserEmailService.sendEmail(user::getEmail)
+            return RegistrationResult(
+                RegistrationResult.Status.ALREADY_EXISTS,
+                registration.emailAddress
+            )
+        }
+//        TODO - implement creation of users as before (ish) for non-SSO
+//             - for SSO users (add domain check) - implement different create user route (no need for password)
+//                    - does this need to come via register page in the same way?
+//                    - is their account just created automatically on first log in or do they still have to register?
```

Comment:
I think ideally it would send them to a pre-filled version of the register page, then if they e.g. already have an account under another email they can realise and try and sort it. If it's easier to just do it in the same request I think that's fine for now though.

And I don't think we need to make them validate their email.

Reply:
Discussion in person concluded:
If people log in via SSO and don't have a delta account yet then we make them a delta account and redirect them to delta logged into that account
If people attempt to register for delta with an email that is one of the SSO required org domains then we redirect them to SSO and once they have logged in treat it as above